src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp

```diff
@@ -0,0 +1,215 @@
+/**
+ * @file q_learning_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is the implementation of QLearning class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_RL_Q_LEARNING_IMPL_HPP
+#define MLPACK_METHODS_RL_Q_LEARNING_IMPL_HPP
+
+#include "q_learning.hpp"
+
+namespace mlpack {
+namespace rl {
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename PolicyType,
+  typename ReplayType
+>
+QLearning<
+  EnvironmentType,
+  NetworkType,
+  OptimizerType,
+  PolicyType,
+  ReplayType
+>::QLearning(NetworkType& network,
+             OptimizerType& optimizer,
+             double discount,
+             PolicyType policy,
+             ReplayType replayMethod,
+             size_t targetNetworkSyncInterval,
+             size_t explorationsSteps,
+             bool doubleQLearning,
+             size_t stepLimit,
+             EnvironmentType environment):
+    learningNetwork(network),
+    optimizer(optimizer),
+    discount(discount),
+    policy(std::move(policy)),
+    replayMethod(std::move(replayMethod)),
+    targetNetworkSyncInterval(targetNetworkSyncInterval),
+    explorationSteps(explorationsSteps),
+    doubleQLearning(doubleQLearning),
+    stepLimit(stepLimit),
+    environment(std::move(environment)),
+    totalSteps(0),
+    deterministic(false)
+{
+  learningNetwork.ResetParameters();
+  targetNetwork = learningNetwork;
+}
+
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename PolicyType,
+  typename ReplayType
+>
+arma::icolvec QLearning<
```

Comment:
Do you think this type should be `arma::Col<size_t>` to match the rest of the codebase better?